improver/calibration/reliability_calibration.py

```diff
@@ -935,6 +941,73 @@ def _update_reliability_table(
         reliability_table.replace_coord(probability_bin_coord)
         return reliability_table
 
+    def _enforce_min_count_and_montonicity(self, rel_table_slice: Cube) -> Cube:
+        """Apply the steps needed to produce a reliability diagram on a single
+        slice of reliability table cube.
+
+        Args:
+            reliability_table_slice:
+                The reliability table slice to be manipulated. The only
+                coordinates expected on this cube are a table_row_index
+                coordinate and corresponding table_row_name coordinate and a
+                probability_bin coordinate.
+        Returns:
+            Processed reliability table slice, with reliability steps applied.
+        """
+        (
+            observation_count,
+            forecast_probability_sum,
+            forecast_count,
+            probability_bin_coord,
+        ) = self._extract_reliability_table_components(rel_table_slice)
+
+        if np.any(forecast_count < self.minimum_forecast_count):
+            (
+                observation_count,
+                forecast_probability_sum,
+                forecast_count,
+                probability_bin_coord,
+            ) = self._combine_undersampled_bins(
+                observation_count,
+                forecast_probability_sum,
+                forecast_count,
+                probability_bin_coord,
+            )
+            rel_table_slice = self._update_reliability_table(
+                rel_table_slice,
+                observation_count,
+                forecast_probability_sum,
+                forecast_count,
+                probability_bin_coord,
+            )
+
+        # If the observation frequency is non-monotonic adjust the
+        # reliability table
```

Comment:
I don't understand this part - why do we expect the observation frequency to be monotonic?

Reply:
Great question, I originally thought it was because since each threshold is calibrated separately, we need to ensure the cdf is monotonic. Now I've read the code better and see that is not the case here since we are looping over threshold.

Would you be able to shed some light @anja-bom or @fionaRust or @gavinevans ?

Reply:
I've just realised that it would be because for a reasonably calibrated model, you would expected observation frequency to increase as the probability bins increased (e.g., for probability bin 0-0.2, you would expected lower obs_count / forecast_count than for probability bin 0.8-1.0).
Of course this would not be the case for a poorly calibrated model, where this monotonicity enforcement kicks in. I am interested in the theory behind this handling of the observation frequency though, I don't think it was in the flowerdew paper?

Reply:
Thanks for the query @lucyleeow. Yes, as you suggest the reason for this manipulation is that we were seeing some reliability diagrams that looked something like this:
 
![Reliability_diagram_example](https://user-images.githubusercontent.com/26933341/156770952-f481e4e7-192d-43fb-9b5d-52e1d1ab5dc5.PNG)

Using a reliability diagram of this form for calibration would mean that high probability forecasts prior to calibration would end up as a lower probability forecast after calibration. Although calibrating using this kind of reliability diagram isn't necessarily "wrong", we were concerned about the use of this kind of diagram leading to spatial artefacts when the calibrated field was visualised on the grid.

For example, the plot below shows a conceptual, highly idealised example of an event with high probabilities at the centre and low probabilities further from the centre. Using the poorly calibrated reliability diagram above might lead to a result, as shown in the central image, where the high probability area prior to calibration has become a lower probability than the surrounding probability areas. Visually this might be a bit odd, so the manipulation ensures that we, at least get something like the image on the right hand side, where the high probability area prior to calibration has become equal to the surrounding probability area. 

![Bullseye_probability_example](https://user-images.githubusercontent.com/26933341/156771636-43accb7f-00f3-4b02-bf72-e73dd0ef4dd9.PNG)

The manipulated reliability diagram would therefore look conceptually something like this:

![Reliability_diagram_manipulated_example](https://user-images.githubusercontent.com/26933341/156778170-97a75ef3-9f5c-43e8-aa07-bb252cfa2468.PNG)

Hopefully that makes some sense conceptually. Happy to have a chat about this next week, if you'd like. 

@fionaRust Anything else to add?

Reply:
Thanks @lucyleeow and @gavinevans, this makes sense now. I think I was confused about the definition of observation frequency, but it's clear now.

Reply:
Thanks for the background @gavinevans, I can understand why that decision was made. Out of curiosity do you have any data on reliability/skill with pre-manipulation calibration and with post-manipulation calibration?

Reply:
No, we haven't compared the reliability/skill of the pre-manipulation calibration vs the post-manipulation calibration beyond ensuring the post-manipulation reliability diagrams looked as desired.